File: Library.java
import java.util.ArrayList;
import java.util.Scanner;

public class Library {
    private ArrayList<User> Userlist = new ArrayList<>();//declarar arreglo
    private ArrayList<Book> Booklist = new ArrayList<>();

    public Library() {
    }

    // CREATE USER
    public void createUser() {
        Scanner sc = new Scanner(System.in);
        System.out.println("Introduce the name: ");
        String name = sc.nextLine();
        System.out.println("Introduce the lastname: ");
        String lastname = sc.nextLine();

        //tiene que haber validación
        if (name.isEmpty() && lastname.isEmpty()) {
            System.out.println("Input invalid...");

        } else {
            System.out.println("Introduce the age");
            int age = sc.nextInt();
            if (age < 6) {
                System.out.println("Not possible to register with that age ");
                return;
            }
            IdGenerator IdGenerator = new IdGenerator();
            String id = IdGenerator.createId();
            User newUser = new User(id, name, lastname, age);
            Userlist.add(newUser);
            System.out.println("User created successfully");
        }
    }
    // REGISTER BOOK
    public void registerBook(){
        Scanner sc = new Scanner(System.in);
        System.out.println("Title : ");
        String title = sc.nextLine();
        System.out.println("Author : ");
        String author = sc.nextLine();
        if (title.isEmpty()&&author.isEmpty()){
            System.out.println("Invalid input, please introduce correctly the data");
            return;
        }
        System.out.println("Publication year : ");
        int year = sc.nextInt();
        if(year>2024){
            System.out.println("Invalid input, please verify");
            return;
        }
        IdGenerator genID = new IdGenerator();
        String id= genID.createBookId();
        Book newBook = new Book(id, title, author, year,true );
        Booklist.add(newBook);
        System.out.println("Book registered succesfully");
    }
    //RENT A BOOK
    public void rentABook(){
        Scanner sc = new Scanner(System.in);
        //listar clientes y seleccionar uno
        System.out.println("Select the client who wants to rent a Book");
        showUsers();
        int option = sc.nextInt();
        option--;
        User Userelected = Userlist.get(option);
        System.out.println("You choosed: "+ " code: "+Userelected.getId()+" | Name: "+Userelected.getName()+" | Lastname: "+Userelected.getLastname()+" | Age: "+Userelected.getAge());
        ////////////////////////////
        System.out.println("Choose the book the client wants to rent: ");
        int id = 0;
        //bookid, title, author, year
        //mostrar solo titulos disponibles
        for(Book book : Booklist){ //objeto loquesea : lista
            if(book.isItsAvailable()) {
                System.out.println("| ID " + (id + 1) + " Id: " + book.getId() + " Title: " + book.getTitle() + " Author: " + book.getAuthor() + " Publication Year: " + book.getYear());
            }
            id++;
        }
        int selection = sc.nextInt();
        selection--;
        //Hasta aquí ya eligió el libro
        Book bookselected = Booklist.get(selection);
        if(!bookselected.isItsAvailable()){
            System.out.println("Sorry, that book isn't available, even it is not shown in the list");
            return;
        }
        System.out.println("You selected:\n  Title: "+bookselected.getTitle()+" | Author: "+bookselected.getAuthor());
        //Userelected.getListRentedBooks().add(bookselected); //Se agrega libro a la lista personal del usuario
        Userlist.get(option).rentABook(bookselected);
        Userlist.get(option).setHasRentedBook(true);
        Booklist.get(selection).setItsAvailable(false); //aquí cambio a que ya no esta disponible
        System.out.println(" >>>Now the rent is authorized \n > The client can take the book ");
    }
    //RETURN BOOK
    public void returnBook(){//si tengo uno para rentar tengo que hacer uno para devolver y que el libro vuelva a estar disponible
        Scanner sc = new Scanner(System.in);
        System.out.println("Select the client who wants to return a book");
        //antes de que sigas escribe aparte el método para mostrar usuarios(encierra el case 2 de showlists e invocalo en este y en el rentabook))
        showUsersHaveRentedBooks();
        int input = sc.nextInt();
        input--;
        User Userelected = Userlist.get(input);
        System.out.println("Which book wants to return?");
        //mostrar lista de libros del ussuario
        seeRentedBooks(Userelected);
    }
    //SEE RETURN BOOKS
    public void seeRentedBooks(User selectedUser){
        Scanner sc = new Scanner(System.in);
        System.out.println("Select the book the client wants to return");
        int count=0;
        for(Book book :selectedUser.getListRentedBooks() ){
            System.out.println("| ID " + (count + 1) + " Id: " +  book.getId()+" Title: " + book.getTitle() + " Author: " + book.getAuthor() + " Publication Year: " + book.getYear());
            count++;
        }
        int selection = sc.nextInt();
        selection--;
        Book bookSelected = selectedUser.getListRentedBooks().get(selection);
        bookSelected.setItsAvailable(true);
        selectedUser.getListRentedBooks().remove(selection);
        System.out.println("The book has been returned");

    }
    //BUY A BOOK
    public void buyABook(){
        Scanner sc = new Scanner(System.in);
        System.out.println("Select the client who wants to buy a Book: ");
        int ID = 0;
        for (User user : Userlist){ //tipoobjeto loquesea: lista
            System.out.println("| ID: "+ (ID+1)+" | code: "+user.getId()+" | Name: "+user.getName()+" | Lastname: "+user.getLastname()+" | Age: "+user.getAge());
            ID++;
        }
        int selection = sc.nextInt();
        selection--;
        User Userelected= Userlist.get(selection);
        System.out.println("Select the book the client wants to buy: ");
        showBooksAvailable();
        int input = sc.nextInt();
        input --;
        Book bookSelected = Booklist.get(input);
        if(!bookSelected.isItsAvailable()){
            System.out.println("Sorry, that book isn't available, even it is not shown in the list");
            return;
        }
        System.out.println("Enter the price: ");
        double price = sc.nextDouble();
        if (price<0){
            System.out.println("Invalid value");
            return;
        }
        System.out.println("The user "+Userelected.getName()+"  "+Userelected.getLastname()+" has bought "+bookSelected.getTitle()+" by $ "+price);

        bookSelected.setItsAvailable(false);
        Userelected.setHasBoughtABook(true);
        Booklist.remove(input);

    }
    //SHOW USER
    public void showUsers (){
        int ID = 0;
        for (User user : Userlist){ //tipoobjeto loquesea: lista
            System.out.println("| ID: "+ (ID+1)+" | code: "+user.getId()+" | Name: "+user.getName()+" | Lastname: "+user.getLastname()+" | Age: "+user.getAge());
            ID++;
        }
    }
    //SHOW USER THAT HAVE RENTED BOOKS
    public void showUsersHaveRentedBooks() {
        int ID = 0;
        for (User user : Userlist) {
            if (user.isHasRentedBook()) {
                System.out.println("|ID: " + (ID + 1) + " | code: " + user.getId() + " | Name: " + user.getName() + " | Lastname: " + user.getLastname() + " | Age: " + user.getAge());
            }
            ID++;
        }
    }
    //SEE ALL BOOKS
    public void showAllBooks(){
        int ID=0;
        for(Book book : Booklist){
            System.out.println(" | ID: "+(ID+1)+" | code: "+book.getId()+" | Title: "+book.getTitle()+" | Author: "+book.getAuthor()+" | Publication year: "+book.getYear());
            ID++;
        }
    }
    //ALL AVAILABLE BOOKS
    public void showBooksAvailable() {
        int ID = 0;
        for (Book availables : Booklist) {
            if (availables.isItsAvailable()) {
                System.out.println("|ID: " + (ID + 1) + " | code: " + availables.getId() + " | Title: " + availables.getTitle() + " | Author: " + availables.getAuthor() + " | Publication year: " + availables.getYear());
            }
            ID++;
        }
    }
    //ALL RENTED BOOKS
    public void showRentedBooks(){
        int ID=0;
        for(Book rentedBooks : Booklist){
            if (!rentedBooks.isItsAvailable()){
                System.out.println("|ID: "+(ID+1)+" | code: "+rentedBooks.getId()+" | Title: "+rentedBooks.getTitle()+"| Author: "+rentedBooks.getAuthor()+" | Publication year: "+rentedBooks.getYear());
            }
            ID++;
        }
    }
    //User THAT HAVE BOUGHT BOOKS
    public void showUserWithRentedBooks(){
        int ID =0;
        for (User user : Userlist){
            if(user.isHasBoughtABook()){
                System.out.println("|ID: "+(ID+1)+" | code: "+user.getId()+" | Name: "+user.getName()+" | Lastname: "+user.getLastname()+ " | Age: "+user.getAge());
            }
            ID++;
        }
    }
    //TEST INFO
    public void testingInfo(){
        ///User
        IdGenerator IdGenerator = new IdGenerator();
        String Id1 = IdGenerator.createId();
        String Id2 = IdGenerator.createId();
        String Id3 = IdGenerator.createId();
        User user1 = new User(Id1, "Alejandro", "Montejano", 20);
        User user2 = new User(Id2, "Atziri", "Mancilla", 27);
        User user3 = new User(Id3, "Jafet", "Santoyo", 27);
        Userlist.add(user1);
        Userlist.add(user2);
        Userlist.add(user3);

        /////Books
        String IdB1 = IdGenerator.createBookId();
        String IdB2 = IdGenerator.createBookId();
        String IdB3 = IdGenerator.createBookId();
        String IdB4 = IdGenerator.createBookId();
        String IdB5 = IdGenerator.createBookId();
        String IdB6 = IdGenerator.createBookId();
        Book book1 = new Book(IdB1, "Muerte en el Nilo", "Agatha Chriestie", 1937,true);
        Book book2 = new Book(IdB2, "Las memorias de Sherlock Holmes", "Arthur Conan Doyle", 1894,true);
        Book book3 = new Book(IdB3, "El pulgar del violinista", "Sam Kean", 2013,true);
        Book book4 = new Book(IdB4, "Momo", "Michael Ende", 1972,true);
        Book book5 = new Book(IdB5, "Salvar el fuego", "Guillermo Arriaga", 2020,true);
        Book book6 = new Book(IdB6, "Ilusiones", "Richard Bach", 1977,true);

        Booklist.add(book1);
        Booklist.add(book2);
        Booklist.add(book3);
        Booklist.add(book4);
        Booklist.add(book5);
        Booklist.add(book6);
    }
}

File: library.py
from user import User
from book import Book
from idgenerator import IdGenerator


class Library:
    def __init__(self):
        self.user_list = []
        self.book_list = []

    def create_user(self):
        name = input("Introduce the name: ")
        lastname = input("Introduce the lastname: ")

        if not name and not lastname:
            print("Input invalid...")
            return

        age = int(input("Introduce the age: "))
        if age < 6:
            print("Not possible to register with that age")
            return

        id_generator = IdGenerator()
        id_ = id_generator.create_id()
        new_user = User(id_, name, lastname, age)
        self.user_list.append(new_user)
        print("User created successfully")

    def register_book(self):
        title = input("Title: ")
        author = input("Author: ")
        if not title and not author:
            print("Invalid input, please introduce correctly the data")
            return

        year = int(input("Publication year: "))
        if year > 2024:
            print("Invalid input, please verify")
            return

        id_generator = IdGenerator()
        id_ = id_generator.create_book_id()
        new_book = Book(id_, title, author, year, True)
        self.book_list.append(new_book)
        print("Book registered successfully")

    def rent_a_book(self):
        print("Select the client who wants to rent a book")
        self.show_users()
        option = int(input()) - 1
        user_selected = self.user_list[option]
        print(f"You chose: code: {user_selected.id}, Name: {user_selected.name}, Lastname: {user_selected.last_name}, Age: {user_selected.age}")

        print("Choose the book the client wants to rent:")
        id_ = 0
        for book in self.book_list:
            if book.is_its_available():
                print(f"| ID {id_ + 1} Id: {book.id}, Title: {book.title}, Author: {book.author}, Publication Year: {book.year}")
            id_ += 1

        selection = int(input()) - 1
        book_selected = self.book_list[selection]
        if not book_selected.is_its_available():
            print("Sorry, that book isn't available, even it is not shown in the list")
            return

        print(f"You selected:\n  Title: {book_selected.title} | Author: {book_selected.author}")
        user_selected.rent_a_book(book_selected)
        user_selected.has_rented_book = True
        self.book_list[selection].its_available = False
        print(">>> Now the rent is authorized\n> The client can take the book")

    def return_book(self):
        print("Select the client who wants to return a book")
        self.show_users_have_rented_books()
        input_ = int(input()) - 1
        user_selected = self.user_list[input_]
        print("Which book wants to return?")
        self.see_rented_books(user_selected)

    def see_rented_books(self, selected_user):
        print("Select the book the client wants to return")
        count = 0
        for book in selected_user.list_rented_books:
            print(f"| ID {count + 1} Id: {book.id} Title: {book.title} Author: {book.author} Publication Year: {book.year}")
            count += 1

        selection = int(input()) - 1
        book_selected = selected_user.list_rented_books[selection]
        book_selected.its_available = True
        selected_user.list_rented_books.remove(book_selected)
        print("The book has been returned")

    def buy_a_book(self):
        print("Select the client who wants to buy a book:")
        ID = 0
        for user in self.user_list:
            print(f"| ID: {ID + 1} | code: {user.id} | Name: {user.name} | Lastname: {user.last_name} | Age: {user.age}")
            ID += 1

        selection = int(input()) - 1
        user_selected = self.user_list[selection]
        print("Select the book the client wants to buy:")
        self.show_books_available()
        input_ = int(input())
        book_selected = self.book_list[input_]
        if not book_selected.is_its_available():
            print("Sorry, that book isn't available, even it is not shown in the list")
            return

        price = float(input("Enter the price: "))
        if price < 0:
            print("Invalid value")
            return

        print(f"The user {user_selected.name} {user_selected.last_name} has bought {book_selected.title} by ${price}")

        user_selected.has_bought_a_book = True
        self.book_list.remove(book_selected)
        book_selected.its_available = False

    def show_users(self):
        for index, user in enumerate(self.user_list):
            print(f"| ID: {index + 1} | code: {user.id} | Name: {user.name} | Lastname: {user.last_name} | Age: {user.age}")

    def show_users_have_rented_books(self):
        for index, user in enumerate(self.user_list):
            if user.has_rented_book:
                print(f"|ID: {index + 1} | code: {user.id} | Name: {user.name} | Lastname: {user.last_name} | Age: {user.age}")

    def show_all_books(self):
        for index, book in enumerate(self.book_list):
            print(f"| ID: {index + 1} | code: {book.id} | Title: {book.title} | Author: {book.author} | Publication year: {book.year}")

    def show_books_available(self):
        for index, available in enumerate(self.book_list):
            if available.is_its_available():
                print(f"|ID: {index + 1} | code: {available.id} | Title: {available.title} | Author: {available.author} | Publication year: {available.year}")

    def show_rented_books(self):
        for index, rented_books in enumerate(self.book_list):
            if not rented_books.is_its_available():
                print(f"|ID: {index + 1} | code: {rented_books.id} | Title: {rented_books.title} | Author: {rented_books.author} | Publication year: {rented_books.year}")

    def show_user_with_rented_books(self):
        for index, user in enumerate(self.user_list):
            if user.has_bought_a_book:
                print(f"|ID: {index + 1} | code: {user.id} | Name: {user.name} | Lastname: {user.last_name} | Age: {user.age}")

    def testing_info(self):
        idgenerator = IdGenerator()
        id_1 = idgenerator.create_id()
        id_2 = idgenerator.create_id()
        id_3 = idgenerator.create_id()
        user_1 = User(id_1, "Alejandro", "Montejano", 20)
        user_2 = User(id_2, "Atziri", "Mancilla", 27)
        user_3 = User(id_3, "Jafet", "Santoyo", 27)
        self.user_list.append(user_1)
        self.user_list.append(user_2)
        self.user_list.append(user_3)
        ##books
        id_b1 = idgenerator.create_book_id()
        id_b2 = idgenerator.create_book_id()
        id_b3 = idgenerator.create_book_id()
        id_b4 = idgenerator.create_book_id()
        id_b5 = idgenerator.create_book_id()
        id_b6 = idgenerator.create_book_id()
        book_1 = Book(id_b1, "Muerte en el Nilo", "Agatha Chriestie", 1937, True)
        book_2 = Book(id_b2, "Las memorias de Sherlock Holmes", "Arthur Conan Doyle", 1894, True)
        book_3 = Book(id_b3, "El pulgar del violinista", "Sam Kean", 2013, True)
        book_4 = Book(id_b4, "Momo", "Michael Ende", 1972, True)
        book_5 = Book(id_b5, "Salvar el fuego", "Guillermo Arriaga", 2020, True)
        book_6 = Book(id_b6, "Ilusiones", "Richard Bach", 1977, True)

        self.book_list.append(book_1)
        self.book_list.append(book_2)
        self.book_list.append(book_3)
        self.book_list.append(book_4)
        self.book_list.append(book_5)
        self.book_list.append(book_6)
